libwnbd/wnbd_ioctl.cpp

```diff
@@ -133,6 +137,315 @@ DWORD WnbdOpenAdapterCMDeviceInstance(PDEVINST DeviceInstance)
+    /* Disallow new mappings so we can remove all current mappings */
+    OptValue.Data.AsBool = FALSE;
+    Status = WnbdSetDrvOpt("NewMappingsAllowed", &OptValue, FALSE);
+    if (Status) {
+        LogError(
+            "WnbdSetDrvOpt Error: %d. Error message: %s", Status, win32_strerror(Status).c_str());
+        goto exit;
+    }
+
+    Status = WnbdList(TempList, &BufferSize);
+    if (!BufferSize) {
+        goto exit;
+    }
+    TempList = (PWNBD_CONNECTION_LIST)calloc(1, BufferSize);
+    if (!TempList) {
+        Status = ERROR_NOT_ENOUGH_MEMORY;
+        LogError(
+            "calloc Error: %d. Error message: %s", Status, win32_strerror(Status).c_str());
+        goto exit;
+    }
+    Status = WnbdList(TempList, &BufferSize);
+    if (Status) {
+        LogError(
+            "WnbdList Error: %d. Error message: %s", Status, win32_strerror(Status).c_str());
+        goto exit;
+    }
+    CHAR* InstanceName;
+    for (ULONG index = 0; index < TempList->Count; index++) {
+        InstanceName = TempList->Connections[index].Properties.InstanceName;
+        WNBD_REMOVE_OPTIONS RemoveOptions = { 0 };
+        RemoveOptions.Flags.HardRemove = TRUE;
+        Status = WnbdRemoveEx(InstanceName, &RemoveOptions);
+        if (Status) {
+            LogError(
+                "Could not remove device: %s. "
+                "Error: %d. Error message: %s",
+                InstanceName, Status, win32_strerror(Status).c_str());
+            goto exit;
+        }
+    }
+
+exit:
+    if (TempList) {
+        free(TempList);
+    }
+    return Status;
+}
+
+DWORD WnbdUninstallDriver(PBOOL RebootRequired)
+{
+    DWORD Status;
+    HDEVINFO AdapterDevHandle = INVALID_HANDLE_VALUE;
+    SP_DEVINFO_DATA DevInfoData;
+    DevInfoData.cbSize = sizeof DevInfoData;
+    Status = FindDevice(&AdapterDevHandle, &DevInfoData);
+    if (ERROR_SUCCESS != Status) {
+        LogError("Driver not installed");
+        return Status;
+    }
+
+    
+    PWNBD_CONNECTION_LIST ConnList = NULL;
+    DWORD err = HardRemoveAllDevices();
+    if (err) {
+        return err;
+    }
+    
+    Status = WnbdRemoveDriver(AdapterDevHandle, &DevInfoData, RebootRequired);
+
+    if (INVALID_HANDLE_VALUE != AdapterDevHandle) {
+        SetupDiDestroyDeviceInfoList(AdapterDevHandle);
+    }
+
+    if (ERROR_SUCCESS != Status) {
+        LogError(
+            "Error: %d. Error message: %s", Status, win32_strerror(Status).c_str());
+    }
+
+    return Status;
+}
+
+DWORD InstallDriver(CHAR* FileNameBuf, HDEVINFO* AdapterDevHandle, PBOOL RebootRequired)
+{
+    CHAR ClassName[MAX_CLASS_NAME_LEN];
+    GUID ClassGuid;
+    SP_DEVINFO_DATA DevInfoData;
+    SP_DEVINSTALL_PARAMS_A InstallParams;
+    DWORD Status = 0;
+
+    if (!SetupDiGetINFClassA(FileNameBuf, &ClassGuid, ClassName, MAX_CLASS_NAME_LEN, 0)) {
+        Status = GetLastError();
+        goto exit;
+    }
+
+    if (INVALID_HANDLE_VALUE == (*AdapterDevHandle = SetupDiCreateDeviceInfoList(&ClassGuid, 0))) {
```

Comment:
It's a bit hard to tell which part of the function registers the driver and which is creating the device. Do you think it would make sense to decouple it, having InstallDriver and CreateWnbdAdapter?